Reader Card Refresh: Stack media vertically and make full width

eoigal opened the pull request:
## Description

We're in the process of updating the look and feel of our Calypso Reader Cards. As a part of this, we'd like to make all media full-width vs left aligned. We'll want to make updates to the following cards:

- Single image
- Two images
- Three images
- Four images
- Video
- Gallery

### Before
![Image](https://user-images.githubusercontent.com/5634774/184982439-5d2d3ec1-441a-4f05-8ca0-0fdc0b942725.png)

### After

#### Single Image
![Image](https://user-images.githubusercontent.com/5634774/184982997-eb81d577-022b-4c83-b527-57461064756c.png)

#### Two image
![Image](https://user-images.githubusercontent.com/5634774/184983147-ab99d5e7-4613-41a6-853a-4577882aaa6b.png)

#### Three image
![Image](https://user-images.githubusercontent.com/5634774/184983214-5ae47825-b043-4428-87c5-14e1bdaf14dc.png)

#### Four image
![Image](https://user-images.githubusercontent.com/5634774/184983280-16e3d583-8628-4af3-9401-376c120bcdbe.png)

#### Video
![Image](https://user-images.githubusercontent.com/5634774/184983431-979dda09-0998-4c50-bdf8-639447f7c329.png)

#### Gallery
![Image](https://user-images.githubusercontent.com/5634774/184983493-978b6411-3854-4be8-a462-c7c6f1e1f5fe.png)

### Changes of note in this PR

* Added a featured images component to show between 1-4 images from a post
* Created flex containers to allow featured images component to display images to match mockups
* Updated the gallery to show 5 - 10 images in full width slideshow format
* Fixed bug in video embeds, where videopress thumbnails were not rendering
* 
*

eoigal committed:
Add featured images component to build flex containers for featured images

eoigal committed:
Update gallery to work as a slideshow

eoigal committed:
fix missing height

eoigal committed:
Add state to handle changing gallery images

eoigal committed:
Sort images by width size and update number of images supported by gallery

eoigal committed:
fix bug with video thumbnails

davemart-in commented:
First review. Noticed a couple of things:

1) Recommended posts (2 posts next to each other) are currently escaping the grid. Both pictures should fit in the same width as the single image:

![CleanShot 2022-09-06 at 10 58 54@2x](https://user-images.githubusercontent.com/5634774/188669035-2ad03c0b-f291-4a0e-b831-427a2656f6f7.png)

2) I did a search for "videopress" and noticed that posts with VideoPress video embeds show up with a black background (no thumbnail):

<img width="1044" alt="CleanShot 2022-09-06 at 11 03 19@2x" src="https://user-images.githubusercontent.com/5634774/188669659-c96cae8a-c9a7-4984-a512-79a93826c6b6.png">

cc @ollierozdarz in case you have any additional feedback. Note: You should be able to click the "Calypso Live" link at the top of this PR to preview these changes:

<img width="944" alt="CleanShot 2022-09-06 at 11 05 17@2x" src="https://user-images.githubusercontent.com/5634774/188670207-2291cdd5-d500-4e14-b89f-34318f6c7519.png">

eoigal commented:
I think there is an issue with Videopress thumbnails (not having one). When I look at search results for youtube the thumbnail of the video is rendering ok. I'll debug further.
<img width="742" alt="Screenshot 2022-09-06 at 18 32 02" src="https://user-images.githubusercontent.com/5560595/188701875-43a696bd-8020-45c5-ba32-8af992d6fb80.png">

eoigal committed:
fix related cards width bug

eoigal commented:
I've fixed the issue with the related posts width and I've reached out to Videopress for more details on why we receive a 403 when trying to access videopress thumbnail URLs - p1662543833203299-slack-C02LT75D3

eoigal committed:
fix video embeds thumbnails

eoigal committed:
add lint changes

eoigal committed:
fix lint issue

roo2 commented on client/blocks/reader-post-card/featured-images.jsx:
```diff
@@ -0,0 +1,89 @@
+
+ReaderFeaturedImages.propTypes = {
+	post: PropTypes.object.isRequired,
+	isDiscover: PropTypes.bool,
```
isDiscover doesn't seem to be used by ReaderFeaturedImages(?)

eoigal replied:
Removed this as not used

roo2 commented on client/blocks/reader-featured-image/index.jsx:
```diff
@@ -2,56 +2,69 @@ import classnames from 'classnames';
+	imageHeight,
 } ) => {
-	if ( imageUrl === undefined ) {
+	const featuredImageUrl = imageUrl || canonicalMedia.src;
```
nit: I think I'd do a null safe check for `canonicalMedia.src` I see the prop is _required_ but I don't think that is really enforced.
```suggestion
	const featuredImageUrl = imageUrl || canonicalMedia?.src;
```

eoigal replied:
Good idea, I'll add this.

eoigal replied:
Good idea, added null safe check

roo2 commented on client/blocks/reader-featured-image/index.jsx:
```diff
@@ -2,56 +2,69 @@ import classnames from 'classnames';
-		<a className={ classNames } href={ href } style={ featuredImageStyle } onClick={ onClick }>
-			{ children }
-		</a>
+		<div style={ divStyle }>
```
nit: I don't understand why the div is now necessary?

eoigal replied:
It isn't necessary. I'll remove it.

eoigal replied:
Not needed, kinda legacy of me trying to get flex containers to work as I'd like

roo2 commented on client/blocks/reader-post-card/gallery.jsx:
```diff
@@ -1,66 +1,111 @@
+		this.setState( { itemSelected: updateItemSelected } );
+	};
+
+	getGalleryWorthyImages = ( post ) => {
```
nit: it looks like this could be pulled out somewhere into a helper function

eoigal replied:
Yeah, good call. I typically leave refactoring until the end but in this case I think ok to move to helper function

roo2 commented on client/blocks/reader-post-card/gallery.jsx:
```diff
@@ -1,66 +1,111 @@
-			};
+	handleClick = ( event ) => {
+		event.preventDefault();
+		let updateItemSelected;
+		updateItemSelected = this.state.itemSelected + 1;
```
nit: This could be one line

eoigal replied:
Yup!

eoigal replied:
👍

roo2 commented on client/blocks/reader-post-card/gallery.jsx:
```diff
@@ -1,66 +1,111 @@
+			const imageUrl = resizeImageUrl( image.src, {
+				w: READER_CONTENT_WIDTH,
+			} );
+			const safeCssUrl = cssSafeUrl( imageUrl );
+			let imageStyle = { background: 'none' };
+			if ( safeCssUrl ) {
+				imageStyle = {
+					backgroundImage: 'url(' + safeCssUrl + ')',
+					backgroundSize: 'cover',
+					backgroundPosition: '50% 50%',
+					backgroundRepeat: 'no-repeat',
+				};
+			}
```
nit: this could potentially also be a helper function shared with reader-featured-images.jsx

eoigal replied:
Agree but prefer composition over early refactoring at this point.

roo2 replied:
that makes sense

roo2 commented on client/blocks/reader-post-card/index.jsx:
```diff
@@ -139,7 +139,7 @@ class ReaderPostCard extends Component {
 			isSeen = post?.is_seen;
 		}
-		const isPhotoPost = !! ( post.display_type & DisplayTypes.PHOTO_ONLY ) && ! compact;
+		const isPhotoPost = false; //!! ( post.display_type & DisplayTypes.PHOTO_ONLY ) && ! compact;
```
commented out code

eoigal replied:
I need to figure out how to handle posts classified as photo posts. This was done just to allow me to see one image featured images.

roo2 commented on client/blocks/reader-post-card/gallery.jsx:
```diff
@@ -1,66 +1,111 @@
+
+			const classes = classnames( {
+				'reader-post-card__gallery-item': true,
+				hide: index !== this.state.itemSelected,
```
I imagine we might want to do some kind of css transition in the future? like fading images in and out rather than just show/hide cc @ollierozdarz @davemart-in

roo2 commented on client/blocks/reader-post-card/gallery.jsx:
```diff
@@ -1,66 +1,111 @@
+};
+
+PostGallery.defaultProps = {
+	onClick: noop,
```
I don't see where `onClick` is used?

eoigal replied:
Not sure why I added that!.. removed and fixed issue with clicking circles.

roo2 commented on client/blocks/reader-post-card/style.scss:
```diff
@@ -408,15 +399,15 @@
 
 	.reader-post-card__post-details {
-		flex: 1 auto;
+		//flex: 1 auto;
```
commented out code

roo2 commented on client/blocks/reader-post-card/style.scss:
```diff
@@ -659,3 +667,26 @@
+	}
+}
+
+/* Create four equal columns that sits next to each other */
```
I think this comment doesn't look related to the CSS below it

eoigal replied:
Removed.

roo2 commented:
It looks like the little circles were supposed to be clickable but they aren't, at least in chrome I'm taken to the article:


https://user-images.githubusercontent.com/22446385/189019833-313bd290-6afb-4841-a432-18a3d630a1df.mov

roo2 commented on client/blocks/reader-post-card/gallery.jsx:
```diff
@@ -1,66 +1,111 @@
+		const images = ( post.images && [ ...post.images ] ) || [];
+
+		return images
+			.sort( ( a, b ) => ( a.width > b.width ? 1 : -1 ) )
```
It looks like this mechanism will show the _largest_ image first, but maybe we should show _the image that appears first in the article_ first (and that is large enough to qualify). What do you think @davemart-in @ollierozdarz ?

eoigal replied:
The sort here was to actually filter out smaller images from the list of images in the post. I suspect that I can rewrite this to avoid sorting (it already sorts by how they are placed in the post afaik) and just filter out smaller/less suitable images that won't suit the gallery view.

eoigal replied:
I've updated this to remove the sort, add a filter to remove duplicates, add a filter to also remove images that are outside an ideal image aspect ratio.

roo2 replied:
EDIT: oops I was looking at an old version 👍

roo2 commented:
hmmm this image looks a bit strange. Maybe we should filter out images that are extremely wide or tall as they're likely not suitable to be resampled for the gallery? cc @davemart-in @ollierozdarz 

I'm sure there are many tall or wide images that would work just fine, but I have a feeling that they're generally less likely to be suitable.

source image:
<img width="694" alt="Screen Shot 2022-09-08 at 12 32 50 pm" src="https://user-images.githubusercontent.com/22446385/189022154-db0c32b2-9493-40f5-9d40-d9614937c89a.png">

in gallery:
<img width="620" alt="Screen Shot 2022-09-08 at 12 44 47 pm" src="https://user-images.githubusercontent.com/22446385/189022147-b9461650-1b32-4a00-94f5-4b42946a8286.png">

roo2 commented:
There seem to be some posts getting duplicate images for some reason. e.g. this feed seems to have the issue on several posts: `/read/feeds/77222666`

<img width="621" alt="Screen Shot 2022-09-08 at 12 52 08 pm" src="https://user-images.githubusercontent.com/22446385/189023250-0d1f1de5-3cfe-4f25-878e-d39e33c2d107.png">

roo2 commented on client/blocks/reader-post-card/gallery.jsx:
```diff
@@ -1,66 +1,111 @@
+				'reader-post-card__gallery-circle': true,
+				'is-selected': index === this.state.itemSelected,
+			} );
+			return <div className={ classes }></div>;
```
I'm getting a warning in the console, I think this is the div that needs a `key` property

eoigal replied:
Fixed

roo2 commented:
it looks like I'm getting a JavaScript error in `/read/conversations`:

<img width="899" alt="Screen Shot 2022-09-08 at 1 04 01 pm" src="https://user-images.githubusercontent.com/22446385/189024744-6e8c42c3-fd1a-4e5f-b8b2-888c4d917143.png">

roo2 commented:
on a positive note, I think this looks amazingly good when the site has nice images selected 😊
<img width="619" alt="Screen Shot 2022-09-08 at 12 55 33 pm" src="https://user-images.githubusercontent.com/22446385/189025111-5084a5f1-ab63-4235-a8ef-784570c371c2.png">

roo2 reviewed (commented):
Looking good, I dived in and did some nitpicking on the code!

eoigal committed:
add filtering by image ratio

eoigal committed:
refactor getting suitable images into utility function - filter duplicates

eoigal committed:
Remove use of isDiscover

eoigal committed:
fix gallery click behaviour

eoigal committed:
tidy css

eoigal commented:
Thanks @roo2 again for review.

I've addressed a few things based on your feedback;

* I've consolidated how we decide if an image is suitable to display as a featured image in either a flex container or gallery
* That function now has 3 filters. One to remove duplicates, one to remove small images and another to remove any images outside an ideal image aspect ratio (still to decide what that ideal is...)
* I also removed the sort from this function so the images should be ordered by where they are placed in post.
* I've removed some cruft from the code that leaked in while building the components

eoigal committed:
fix missing post prop

eoigal committed:
refactor gallery to only re-render images when clicked on

eoigal committed:
Enable photo posts again but increase height to match featured image

davemart-in commented:
@eoigal I did another quick review. When you're ready, I'd say let's just merge this back into #66650 and then we can do any additional bug fixes or fine tuning directly in that PR.

davemart-in committed:
Added animation transition to gallery

davemart-in commented:
@eoigal I went ahead and [updated the transition animation](https://github.com/Automattic/wp-calypso/pull/67387/commits/8fe67c58af43f02687851ed9798aff424f091ae1) when going from one gallery image to another.

### Before

![CleanShot 2022-09-09 at 11 42 16](https://user-images.githubusercontent.com/5634774/189389370-a64a7d03-b095-47c9-8244-a09c24779466.gif)

### After

![CleanShot 2022-09-09 at 11 44 05](https://user-images.githubusercontent.com/5634774/189389590-f774d38b-0c38-4fcb-9293-86ab38eeb215.gif)

roo2 commented on client/blocks/reader-post-card/style.scss:
```diff
@@ -595,13 +595,19 @@
 	border: 1px solid var( --color-neutral-0 );
 	cursor: pointer;
 	flex: 1;
+	height: 300px;
```
I haven't tested this but will the `reader-post-card__gallery-item` always be 598x300, even on mobile?

davemart-in replied:
Updated this in https://github.com/Automattic/wp-calypso/pull/67387/commits/2b98ad641eced13ce7b771c1b937ab92d48184f6

roo2 commented:
nice work @davemart-in it's those little touches 👨‍🍳

roo2 commented:
Also an image on the `/read/conversations` tab looks mis-sized
<img width="739" alt="Screen Shot 2022-09-12 at 10 21 19 pm" src="https://user-images.githubusercontent.com/22446385/189652033-ce118cf4-8042-4768-914a-4fbe717ecc7b.png">

davemart-in commented:
I'll take a look at both of those this morning. Thanks for highlighting @roo2!

davemart-in committed:
Fix gallery issue where images disappear on mobile. Addresses issue mentioned here: https://github.com/Automattic/wp-calypso/pull/67387#issuecomment-1243658345

davemart-in committed:
Fixed issue with malformed images in conversations. Addresses https://github.com/Automattic/wp-calypso/pull/67387#issuecomment-1243659873

davemart-in commented:
**Updates**

1) In [this commit](https://github.com/Automattic/wp-calypso/pull/67387/commits/2b98ad641eced13ce7b771c1b937ab92d48184f6) I fixed the gallery issue where images disappear on mobile, addressing [Eru's issue reported here](https://github.com/Automattic/wp-calypso/pull/67387#issuecomment-1243658345). Here's a preview of the fixed state:

![CleanShot 2022-09-13 at 13 41 30](https://user-images.githubusercontent.com/5634774/189973026-861a9859-c5e6-45b8-b3cd-e103fe9c26d8.gif)

2) In [this commit](https://github.com/Automattic/wp-calypso/pull/67387/commits/f304574886463276e76414e10df0b058408ba328) I fixed the issue with malformed images in conversations, which addresses [this issue posted by Eru](https://github.com/Automattic/wp-calypso/pull/67387#issuecomment-1243659873). Here's a preview:

<img width="643" alt="CleanShot 2022-09-13 at 13 53 18@2x" src="https://user-images.githubusercontent.com/5634774/189975060-611c373f-354c-421d-b956-22c9406990cf.png">

3) @roo2 in [this issue](https://github.com/Automattic/wp-calypso/pull/67387#issuecomment-1243666916) you mentioned a slight jump. I was unable to replicate this. I may have fixed it with the changes above. Mind double checking?

cc @eoigal
